### PR TITLE
STITCH-1040 Improve semantics of multiple logins

### PR DESCRIPTION
**Pull request by adamchel**
In this change I now keep track of the auth provider used to log into a session in a new UserDefaults key, I exposed that in the StitchClient, and I updated `login()` to use the expected semantics.

**Drive-by changes:**
- Updated the enum of AuthProviderTypes to be `internal` rather than `private` so I could compare the types in the `StitchClient`. 
- Changed the hardcoded types in each of the the auth providers to the raw value of the enum type

### CODE REVIEW

File: StitchCore/StitchCore/Source/Core/StitchHTTPClient.swift [hunk 1]
```diff
@@ -102,6 +106,8 @@ internal class StitchHTTPClient {
         try deleteToken(withKey: Consts.AuthRefreshTokenKey)
         try deleteToken(withKey: Consts.IsLoggedInUDKey)
```
Comment (adamchel) [inline StitchCore/StitchCore/Source/Core/StitchHTTPClient.swift]: I think we should be doing `userDefaults?.removeObject(forKey: Consts.IsLoggedInUDKey)` here because with `deleteToken` it will try to delete the key from the keychain when not using the simulator. Should I make this drive-by change or is this here intentionally?

File: StitchCore/StitchCore/Source/Core/StitchClient.swift [hunk 2]
```diff
@@ -126,6 +129,11 @@ public class StitchClient: StitchClientType {
     public var isAuthenticated: Bool {
         return self.httpClient.isAuthenticated
     }
+    
+    // Returns the type of the provider used to log into the current session. nil if not authenticated
+    public var loggedInProviderType: String? {
+        return self.httpClient.loggedInProviderType
```
Comment (edaniels) [inline StitchCore/StitchCore/Source/Core/StitchClient.swift]: It seems odd that `loggedInProviderType` is exposed via the HTTP client

File: StitchCore/StitchCore/Source/Core/StitchClient.swift [hunk 3]
```diff
@@ -268,27 +276,41 @@ public class StitchClient: StitchClientType {
     public func login(withProvider provider: AuthProvider) -> Promise<UserId> {
         self.authProvider = provider
 
-        if isAuthenticated, let auth = auth {
-            printLog(.info, text: "Already logged in, using cached token.")
-            return Promise.init(value: auth.userId)
+        func doLoginRequest() -> Promise<UserId> {
+            return httpClient.doRequest {
```
Comment (edaniels) [inline StitchCore/StitchCore/Source/Core/StitchClient.swift]: [style] This is more for my education, but how much do we personally like the usage of `${0..n}` as opposed to just specifying the parameter in the function closure?

File: StitchCore/StitchCore/Source/Core/StitchClient.swift [hunk 4]
```diff
@@ -268,27 +276,41 @@ public class StitchClient: StitchClientType {
...
-            strongSelf.onLogin()
-            return authInfo.userId
+        if isAuthenticated, let auth = auth {
+            if provider.type == AuthProviderTypes.anonymous.rawValue &&
+                loggedInProviderType == AuthProviderTypes.anonymous.rawValue {
+                printLog(.info, text: "Already logged in as anonymous user, using cached token.")
+                return Promise.init(value: auth.userId)
+            }
+
+            // Using a different provider, log outand then perform login.
```
Comment (edaniels) [inline StitchCore/StitchCore/Source/Core/StitchClient.swift]: [nit] split `outand`

File: StitchCore/StitchCore/Source/Core/StitchClient.swift [hunk 5]
```diff
@@ -268,27 +276,41 @@ public class StitchClient: StitchClientType {
...
-            guard let strongSelf = self else { throw StitchError.clientReleased }
-            let authInfo = try JSONDecoder().decode(AuthInfo.self,
-                                                    from: JSONSerialization.data(withJSONObject: any))
-            strongSelf.httpClient.authInfo = authInfo
-            strongSelf._auth = Auth(stitchClient: strongSelf,
-                                    stitchHttpClient: strongSelf.httpClient,
-                                    userId: authInfo.userId)
-            strongSelf.onLogin()
-            return authInfo.userId
+        if isAuthenticated, let auth = auth {
```
Comment (edaniels) [inline StitchCore/StitchCore/Source/Core/StitchClient.swift]: [style] feels odd to scope an `auth` by assigning it from `auth`

File: StitchCore/StitchCore/Source/Core/StitchClient.swift [hunk 6]
```diff
@@ -268,27 +276,41 @@ public class StitchClient: StitchClientType {
...
+            if provider.type == AuthProviderTypes.anonymous.rawValue &&
+                loggedInProviderType == AuthProviderTypes.anonymous.rawValue {
```
Comment (edaniels) [inline StitchCore/StitchCore/Source/Core/StitchClient.swift]: [note] This will have the effect of logging out an anonymous user since we aren't migrating. This results in a loss of an anonymous account which is okay during our stage of development. We should document it though.

File: StitchCore/StitchCore/Source/Core/StitchClient.swift [hunk 7]
```diff
@@ -268,27 +276,41 @@ public class StitchClient: StitchClientType {
...
+                printLog(.info, text: "Already logged in as anonymous user, using cached token.")
+                return Promise.init(value: auth.userId)
+            }
+
+            // Using a different provider, log outand then perform login.
+            printLog(.info, text: "Already logged in, logging out of existing session.")
+            return self.logout().then {
```
Comment (edaniels) [inline StitchCore/StitchCore/Source/Core/StitchClient.swift]: [minor] If logout fails, we still want to log in. See how the JS SDK does this in the logout method by handling both resolve and reject cases.

File: StitchCore/StitchCore/Source/Core/StitchClient.swift (hunk 5 above)
Comment (edaniels) [inline StitchCore/StitchCore/Source/Core/StitchClient.swift]: [nit] Easier to return early in a small block of code checking `if !isAuthenticated`.

[style?] In fact, it must be the case that auth is not `nil`, which I think means we can use the force unwrap expression `!auth.userId`. @jsflax, as the owners of the state of `auth`, is it fair to say we want a runtime error to happen if `auth` is `nil` when `isAuthenticated` is true or is it too dangerous.

File: StitchCore/StitchCoreTests/AuthTests.swift [hunk 8]
```diff
@@ -115,4 +115,42 @@ class AuthTests: XCTestCase {
 
         wait(for: [exp], timeout: 10)
     }
+    
+    func testMultipleLoginSemantics() throws {
```
Comment (edaniels) [inline StitchCore/StitchCoreTests/AuthTests.swift]: I suppose there isn't really a good way to test logout failing with our current setup unless a mocking library exists.

File: StitchCore/StitchCore/Source/Core/StitchClient.swift (hunk 2 above)
Comment (adamchel) [reply StitchCore/StitchCore/Source/Core/StitchClient.swift]: I did it this way since the HTTP Client is where Auth is actually cleared (and thus the lowest-level part of the client that touches the UserDefaults, but I guess  since nothing in the HTTP Client is accessing `loggedInProviderType` I can move it to `StitchClient`

File: StitchCore/StitchCore/Source/Core/StitchClient.swift (hunk 5 above)
Comment (adamchel) [reply StitchCore/StitchCore/Source/Core/StitchClient.swift]: I think the intention here is to ensure that auth isn't nil and thus to avoid doing `auth!`in the if block.

File: StitchCore/StitchCore/Source/Core/StitchClient.swift (hunk 7 above)
Comment (adamchel) [reply StitchCore/StitchCore/Source/Core/StitchClient.swift]: Done.

That being said, I'm not sure this is the behavior currently in the JS SDK, because if you look at this line: https://github.com/mongodb/stitch-js-sdk/blob/master/src/client.js#L171

`authenticateFn` is only being called on the resolve case of the `logout` promise. The `logout` method handling both resolve and reject in the JS SDK is for clearing auth in storage, not for authenticating.

File: StitchCore/StitchCore/Source/Core/StitchClient.swift (hunk 6 above)
Comment (adamchel) [reply StitchCore/StitchCore/Source/Core/StitchClient.swift]: Would we just put that in release notes somewhere or do you want me to add a comment somewhere?

File: StitchCore/StitchCore/Source/Core/StitchClient.swift (hunk 5 above)
Comment (adamchel) [reply StitchCore/StitchCore/Source/Core/StitchClient.swift]: Holding off on @jsflax 's review before making changes here. But if we agree that a runtime error is okay then I can change this to do a `guard isAuthenticated else { return doLoginRequest() }` block instead of the big `if` block.

File: StitchCore/StitchCore/Source/Core/StitchClient.swift (hunk 3 above)
Comment (adamchel) [reply StitchCore/StitchCore/Source/Core/StitchClient.swift]: I'm usually for explicitly specifying parameters for the sake of readability but I'll let @jsflax decide since he's more familiar with what's normal in Swift.

File: StitchCore/StitchCore/Source/Core/StitchClient.swift (hunk 6 above)
Comment (edaniels) [reply StitchCore/StitchCore/Source/Core/StitchClient.swift]: Release notes

**Comment by adamchel** [comment]
Per discussion with Eric, I made another drive-by change which makes the `logout()` method on the StitchClient semantically identical to `logout()` in the JS SDK. 

Essentially, any error in the logout request—which was previously being thrown as an error to the user—is now being ignored (since the logout request doesn't do anything too important that the user needs to know about). 

The only time an error will be thrown from `logout()` is if there is an issue with clearing the local auth storage, which is really the only important thing `logout()` is doing from an SDK user's perspective.

File: StitchCore/StitchCore/Source/Auth/Providers/Anonymous/AnonymousAuthProvider.swift [hunk 9]
```diff
@@ -3,7 +3,7 @@ import ExtendedJson
 /// AnonymousAuthProvider provides a way to authenticate anonymously.
 public struct AnonymousAuthProvider: AuthProvider {
     /// The authentication type for anonymous login.
-    public let type: String = "anon-user"
+    public let type: String = AuthProviderTypes.anonymous.rawValue
```
Comment (jsflax) [inline StitchCore/StitchCore/Source/Auth/Providers/Anonymous/AnonymousAuthProvider.swift]: [minor] All of these can probably be `AuthProviderTypes` now. I'd like to clean this `protocol` up a bit in general, but there's no reason we should be declaring the rawValue for each of these types.

File: StitchCore/StitchCore/Source/Auth/Providers/ApiKey/ApiKeyAuthProvider.swift [hunk 10]
```diff
@@ -12,7 +12,7 @@ import ExtendedJson
 public struct ApiKeyAuthProvider: AuthProvider {
     private let key: String
 
-    public var type: String = "api-key"
+    public var type: String = AuthProviderTypes.apiKey.rawValue
```
Comment (jsflax) [inline StitchCore/StitchCore/Source/Auth/Providers/ApiKey/ApiKeyAuthProvider.swift]: [minor] Same as above.

File: StitchCore/StitchCore/Source/Auth/Providers/Custom/CustomAuthProvider.swift [hunk 11]
```diff
@@ -4,7 +4,7 @@ import ExtendedJson
 // CustomAuthProvider is a special case that does not
 // follow previous protocols.
 public struct CustomAuthProvider: AuthProvider {
-    public var type: String = "custom-token"
+    public var type: String = AuthProviderTypes.custom.rawValue
```
Comment (jsflax) [inline StitchCore/StitchCore/Source/Auth/Providers/Custom/CustomAuthProvider.swift]: [minor] Same as above.

File: StitchCore/StitchCore/Source/Auth/Providers/EmailPassword/EmailPasswordAuthProvider.swift [hunk 12]
```diff
@@ -5,7 +5,7 @@ import ExtendedJson
 public struct EmailPasswordAuthProvider: AuthProvider {
     /// The authentication type for email/pass login.
     public var type: String {
-        return "local-userpass"
+        return AuthProviderTypes.emailPass.rawValue
```
Comment (jsflax) [inline StitchCore/StitchCore/Source/Auth/Providers/EmailPassword/EmailPasswordAuthProvider.swift]: [minor][nit] Same as above, and this doesn't need to be a computed property.

File: StitchCore/StitchCore/Source/Auth/Providers/Facebook/FacebookAuthProvider.swift [hunk 13]
```diff
@@ -5,7 +5,7 @@ import ExtendedJson
 public struct FacebookAuthProvider: AuthProvider {
     /// The authentication type for facebook login.
     public var type: String {
-        return "oauth2-facebook"
+        return AuthProviderTypes.facebook.rawValue
```
Comment (jsflax) [inline StitchCore/StitchCore/Source/Auth/Providers/Facebook/FacebookAuthProvider.swift]: [minor][nit] Same as above, and this doesn't need to be a computed property.

File: StitchCore/StitchCore/Source/Auth/Providers/Google/GoogleAuthProvider.swift [hunk 14]
```diff
@@ -5,7 +5,7 @@ import ExtendedJson
 public struct GoogleAuthProvider: AuthProvider {
     /// The authentication type for google login.
     public var type: String {
-        return "oauth2-google"
+        return AuthProviderTypes.google.rawValue
```
Comment (jsflax) [inline StitchCore/StitchCore/Source/Auth/Providers/Google/GoogleAuthProvider.swift]: [minor][nit] Same as above, and this doesn't need to be a computed property.

File: StitchCore/StitchCore/Source/Core/StitchClient.swift (hunk 5 above)
Comment (jsflax) [reply StitchCore/StitchCore/Source/Core/StitchClient.swift]: Conditional bindings also allow inner properties of structures, so we could just strip out `userId`. Adam's idea of using a `guard` here also works.

File: StitchCore/StitchCoreTests/AuthTests.swift (hunk 8 above)
Comment (jsflax) [reply StitchCore/StitchCoreTests/AuthTests.swift]: Correct ... I may just push the Admin client forward sooner than later at this point. These tests are only going to be flimsy until then.

This change will come alongside a mocking framework (which we'll have to pick) and a matcher framework (https://github.com/Quick/Nimble) which will make our testing much more robust.

**Comment by adamchel** [comment]
Just went through Jason’s comments. In addition to the requested changes:

- `AuthProviderTypes` is now a `public` enum (was originally `private`, then `internal` for my first pass)
- I added an informational log message that happens when the request  in `logout()` fails.
- The computed `loggedInProviderType` property on `StitchClient` now returns an `Optional(AuthProviderTypes)` rather than an `Optional(String)`
